mirage-router/unikernel.ml

```diff
@@ -0,0 +1,271 @@
+                        Ipv4_packet.pp hdr);
+          Error `Drop
+        end else
+          Error (`Icmp (icmp_err Time_exceeded))
+      else begin
+        Logs.warn (fun m -> m "packet not first fragment %a, TTL exceeded"
+                        Ipv4_packet.pp hdr);
+        Error `Drop
+      end
+    else if hdr.Ipv4_packet.off land 0x4000 = 0x4000 &&
+            Cstruct.len payload > mtu then (* don't fragment set and would fragment *)
+      if is_first_fragment then
+        if is_icmp then begin
+          Logs.warn (fun m -> m "received ICMP %a with don't fragment, but would fragmentation"
+                        Ipv4_packet.pp hdr);
+          Error `Drop
+        end else
+          let code = unreachable_reason_to_int Would_fragment
+          and subheader = Next_hop_mtu mtu
+          in
+          Error (`Icmp (icmp_err ~subheader ~code Destination_unreachable))
+      else begin
+        Logs.warn (fun m -> m "packet not first fragment %a, would fragment"
+                      Ipv4_packet.pp hdr);
+        Error `Drop
+      end
+    else
+      let hdr = { hdr with Ipv4_packet.ttl = pred hdr.Ipv4_packet.ttl } in
+      Ok hdr
+
+  type t = {
+    ovpn : O.t ;
+    mutable ovpn_fragments : Fragments.Cache.t ;
+    private_ip : I.t ;
+    mutable private_fragments : Fragments.Cache.t ;
+  }
+
+  let private_recv net eth arp t =
+    let our_listen =
+      E.input
+        ~arpv4:(A.input arp)
+        ~ipv4:(fun _ -> Logs.warn (fun m -> m "ignoring IPv4"); Lwt.return_unit)
+        ~ipv6:(fun _ -> Logs.warn (fun m -> m "ignoring IPv6"); Lwt.return_unit)
+        eth
+    and forward_packet ip_hdr payload =
+      let c, pkt =
+        Fragments.process t.private_fragments (M.elapsed_ns ()) ip_hdr payload
+      in
+      t.private_fragments <- c;
+      Logs.info (fun m -> m "%B forwarding packet %a"
+                    (match pkt with None -> false | Some _ -> true)
+                    Ipv4_packet.pp ip_hdr);
+      match pkt with
+      | None -> Lwt.return_unit
+      | Some (hdr, pay) ->
+        let pay_mtu = O.mtu t.ovpn - 20 in
+        match forward_or_reject hdr pay pay_mtu with
+        | Ok hdr ->
+          let hdr, fst, rest =
+            if Cstruct.len pay > pay_mtu then
+              let fst, rest = Cstruct.split pay pay_mtu in
+              let hdr = { hdr with Ipv4_packet.off = 0x2000 } in
+              hdr, fst, Fragments.fragment ~mtu:pay_mtu hdr rest
+            else
+              hdr, pay, []
+          in
+          let hdr_cs =
+            Ipv4_packet.Marshal.make_cstruct ~payload_len:(Cstruct.len fst) hdr
+          in
+          let write_one data = O.write t.ovpn data >|= fun _ -> () in
+          write_one (Cstruct.append hdr_cs pay) >>= fun () ->
+          Lwt_list.iter_s write_one rest
+        | Error (`Icmp payload) ->
+          Lwt.async (fun () ->
+              I.write t.private_ip ~ttl:64 hdr.Ipv4_packet.src `ICMP
```

Comment:
Since we're not using path discovery, any reason this is not the max / same as here https://github.com/roburio/openvpn/pull/37/files#diff-f9e65b01a04809253309112aed0616beR234 ?

Reply:
done in d1c9579